### DISCUSSION TITLE
QskLinearBox‘problem

### DISCUSSION THREAD

**Discussion by hestrro** [uwerat/qskinny#494] [Q&A] [answered]
When I add three buttons to the QskLinearBox, layout them vertically, and set the spacing to 0
class Content:public QskLinearBox
{
public:
    Content(QQuickItem *parent=nullptr):QskLinearBox(parent)
    {
        this->setDefaultAlignment(Qt::AlignTop);
        setOrientation(Qt::Vertical);
        setSpacing(0);
    }
};
But it's widely spaced：
![image](https://github.com/user-attachments/assets/7c18639d-920b-4cb6-bb58-a4a6037a7e5d)
This is my attempt to add a stretch factor at index 10 in the constructor：
class Content:public QskLinearBox
{
public:
    Content(QQuickItem *parent=nullptr):QskLinearBox(parent)
    {
        this->setDefaultAlignment(Qt::AlignTop);
        setOrientation(Qt::Vertical);
        setSpacing(0);
        insertStretch(10,1);
    }

};
The buttons added later seem to be added after index=10
![image](https://github.com/user-attachments/assets/c493db64-249a-42d5-9f16-3a85f4f8edf6)

Why does setting setDefaultAlignment(Qt::AlignTop); and setSpacing(0); in a vertical layout seem to have no effect, but it works when it is a horizontal layout?

**Comment by hestrro** [comment]
How can I make the first three buttons at the top? These three buttons are added through QskLinearbox's additem after the constructor ends


Uploading 20241213_155103.mp4…

**Comment by hestrro** [comment]
https://github.com/user-attachments/assets/50e4ea1d-e08a-40a9-ab1f-70cac606436c

**Comment by uwerat** [comment]
The alignment is about how to align a control into its slot, not about aligning the slot.  In your case the layout can't distribute all available space to the controls as they refuse ( according to their size policy ) to accept more.

What to do with this extra spacing can be set by setExtraSpacingAt. As you want to have it at the bottom only you have to set:

    box->setExtraSpacingAt( Qt::BottomEdge );

**Comment by uwerat** [comment]
Using a stretch would work as well, but the index ( 10 ) is wrong: It has to be: addStretch( 1 ) or insertStretch( -1, 1 );

**Reply by hestrro** [reply]
![image](https://github.com/user-attachments/assets/106ebb0c-d98e-416a-a340-25ccf79bbd6d)
The button added later is at the bottom, I want it to be at the top
class Content:public QskLinearBox
{
public:
    Content(QQuickItem *parent=nullptr):QskLinearBox(parent)
    {
        this->setDefaultAlignment(Qt::AlignTop|Qt::AlignLeft);
       setOrientation(Qt::Vertical);
        //setOrientation(Qt::Horizontal);
        setSpacing(0);
       //insertStretch(-1,1);
       setExtraSpacingAt( Qt::BottomEdge );
    }

};
setExtraSpacingAt(Qt::BottomEdge); The situation does not change：
![image](https://github.com/user-attachments/assets/2b1dc4a2-c5d4-4501-ab9f-614b7e15ba7f)

**Reply by hestrro** [reply]
I added the colored boxes together with the buttons, but I hid them via box->setVisible(false);. This should not affect the QskLinearBox.

**Reply by uwerat** [reply]
Strip your example to the very minimum and upload the complete code.

**Comment by hestrro** [comment]
[code.txt](https://github.com/user-attachments/files/18123546/code.txt)

**Reply by uwerat** [reply, marked as answer]
By minimal I mean really minimal and of course compilable. If you want me to look at your code I don't want to spend my time on steps that are unrelated to the problem. ( won't do it next time ! )

Now to your layout issue:

a) remove the setFixedHeight( 200 ) lines
b) do not allow ToolBox to gain extra space ( toolBox->setSizePolicy( Qt::Vertical, QskSizePolicy::Fixed )
c) now Content can't give the extra space to any of its children
d) content->setExtraSpacingAt( Qt::BottomEdge ) - then the extra space goes to the bottom and the buttons are on top

The actual problem of your code is b)

2 notes:

- use QGuiApplication instead of QApplication and remove Qt/Widgets from your dependences ( Qt/Quick does not depend on it )
- QskControl::setBackgroundColor() often helps to analyze layout issues ( for QskBox you have to use setPanel( false ) - otherwise the background is below the panel )
- I recommend to copy the code of  SkinnyShortcut::showBackground into your application and connect it to some shortcut. In the Qskinny examples the shortcut is CTRL-B. It also helps a lot with layout issues

**Reply by hestrro** [reply]
Thank you, I don't want you to waste your time looking at the code. Sorry. I am not very good at using GitHub yet. I will learn how to use GitHub later

**Reply by hestrro** [reply]
QskSetup::setUpdateFlag( QskItem::DebugForceBackground, true );
    w.setCustomRenderMode("overdraw");
I have seen the function setCustomRenderMode("overdraw"); before but don not know how to use it.
![image](https://github.com/user-attachments/assets/60e3558e-4958-47d4-bb7d-ddba09fcfda1)

**Reply by hestrro** [reply]
Sometimes I need to use qweight controls, such asQFileDialog::getOpenFileName
![image](https://github.com/user-attachments/assets/4a1a16dc-9f76-4a98-8a23-423ac6ac4588)
So use QApplication
